src/block_service/hdfs/hdfs_service.cpp

```diff
@@ -386,7 +390,10 @@ error_code hdfs_file_object::read_data_in_batches(uint64_t start_pos,
     uint64_t read_size = 0;
     bool read_success = true;
     while (cur_pos < start_pos + data_length) {
+        auto rate = FLAGS_hdfs_read_limit_rate << 20;
         read_size = std::min(start_pos + data_length - cur_pos, FLAGS_hdfs_read_batch_size_bytes);
+        _service->_read_token_bucket->consumeWithBorrowAndWait(read_size, rate, 2 * rate);
```

Comment:
I think you should add a judgement here to see if this function call of `consumeWithBorrowAndWait` is succeed.